File: packages/cli/src/commands/yank.ts
import { Command } from 'commander';
import { createInterface } from 'node:readline/promises';
import { icons, c, log, logJson, logError, withSpinner, getCurrentMode } from '../lib/output.js';
import { loadConfig } from '../lib/config.js';
import { createApiClient, ApiClientError } from '../lib/api-client.js';

/**
 * Parse a "name@version" string into { name, version }.
 */
const parseNameVersion = (specifier: string): { name: string; version: string } | null => {
  const atIdx = specifier.lastIndexOf('@');
  if (atIdx <= 0) return null;

  const name = specifier.slice(0, atIdx);
  const version = specifier.slice(atIdx + 1);

  if (!name || !version) return null;
  return { name, version };
};

/**
 * Prompt the user for confirmation.
 */
const confirmAction = async (message: string): Promise<boolean> => {
  const rl = createInterface({ input: process.stdin, output: process.stdout });
  try {
    const answer = await rl.question(`  ${message} (y/N) `);
    return answer.trim().toLowerCase() === 'y';
  } finally {
    rl.close();
  }
};

export const registerYankCommand = (program: Command): void => {
  program
    .command('yank <specifier>')
    .description('Yank a specific version from the registry')
    .option('--reason <reason>', 'Reason for yanking')
    .option('--force', 'Skip confirmation prompt')
    .action(async (specifier: string, opts: { reason?: string; force?: boolean }) => {
      const mode = getCurrentMode();

      // 1. Check auth
      const config = loadConfig();
      if (!config.token) {
        logError(
          'Not authenticated',
          'You must be logged in to yank versions.',
          `Run ${c.cmd('spm login')} to authenticate.`,
        );
        process.exitCode = 1;
        return;
      }

      // 2. Parse name@version
      const parsed = parseNameVersion(specifier);
      if (!parsed) {
        logError(
          'Invalid specifier',
          `Expected format: ${c.cmd('name@version')}`,
          `Example: ${c.cmd('spm yank my-skill@1.2.3')}`,
        );
        process.exitCode = 1;
        return;
      }

      const { name, version } = parsed;

      // 3. Confirm
      if (!opts.force) {
        const confirmed = await confirmAction(
          `Yank ${c.name(name)}@${c.version(version)}? This will prevent new installs of this version.`,
        );
        if (!confirmed) {
          log(`${icons.info} Cancelled.`);
          return;
        }
      }

      // 4. Call API
      const api = createApiClient(config);

      try {
        await withSpinner(`Yanking ${name}@${version}...`, () =>
          api.yankVersion(name, version, opts.reason),
        );

        log(`${icons.success} Yanked ${c.name(name)}@${c.version(version)}`);

        if (mode === 'json') {
          logJson({
            command: 'yank',
            status: 'success',
            name,
            version,
            reason: opts.reason,
          });
        }
      } catch (err) {
        if (err instanceof ApiClientError) {
          if (err.status === 404) {
            logError(
              'Version not found',
              `${name}@${version} does not exist in the registry.`,
              `Run ${c.cmd(`spm info ${name}`)} to see available versions.`,
            );
          } else if (err.status === 401) {
            logError(
              'Authentication failed',
              'Your token may have expired.',
              `Run ${c.cmd('spm login')} to re-authenticate.`,
            );
          } else if (err.status === 403) {
            logError('Permission denied', 'You can only yank versions of skills you own.');
          } else {
            logError('Yank failed', err.apiError.message, err.apiError.suggestion);
          }
        } else {
          logError('Yank failed', err instanceof Error ? err.message : String(err));
        }
        process.exitCode = 1;
      }
    });
};

File: cli-go/cmd/yank.go
package cmd

import (
	"fmt"
	"strings"

	"github.com/skillpkg/spm/internal/api"
	"github.com/skillpkg/spm/internal/output"
	"github.com/spf13/cobra"
)

var yankReason string

var yankCmd = &cobra.Command{
	Use:   "yank <name@version>",
	Short: "Yank a specific version from the registry",
	Long:  "Remove a specific skill version from the registry so it can no longer be installed. Existing installs are unaffected.",
	Args:  cobra.ExactArgs(1),
	RunE:  runYank,
}

func init() {
	yankCmd.Flags().StringVar(&yankReason, "reason", "", "Reason for yanking")
	rootCmd.AddCommand(yankCmd)
}

// parseNameVersion splits a "name@version" specifier.
// Handles scoped packages like @scope/name@version.
func parseNameVersion(specifier string) (name, version string, ok bool) {
	// Find the last '@' that is not at position 0 (scoped package prefix)
	atIdx := strings.LastIndex(specifier, "@")
	if atIdx <= 0 {
		return "", "", false
	}
	name = specifier[:atIdx]
	version = specifier[atIdx+1:]
	if name == "" || version == "" {
		return "", "", false
	}
	return name, version, true
}

func runYank(_ *cobra.Command, args []string) error {
	specifier := args[0]

	// 1. Check auth
	if Cfg.Token == "" {
		if Out.Mode == output.ModeJSON {
			return Out.LogJSON(map[string]any{
				"command": "yank",
				"status":  "error",
				"error":   "not_logged_in",
			})
		}
		Out.LogError("Not authenticated. Run %s to log in.", output.Cyan("spm login"))
		return fmt.Errorf("not logged in")
	}

	// 2. Parse name@version
	name, version, ok := parseNameVersion(specifier)
	if !ok {
		if Out.Mode == output.ModeJSON {
			return Out.LogJSON(map[string]any{
				"command": "yank",
				"status":  "error",
				"error":   "invalid_specifier",
				"message": fmt.Sprintf("expected format: name@version, got: %s", specifier),
			})
		}
		Out.LogError("Invalid specifier: expected name@version (e.g. my-skill@1.2.3)")
		return fmt.Errorf("invalid specifier: %s", specifier)
	}

	// 3. Call API
	client := api.NewClient(Cfg.RegistryURL(), Cfg.Token)

	sp := Out.StartSpinner(fmt.Sprintf("Yanking %s@%s...", name, version))
	resp, err := client.Yank(name, version, yankReason)
	output.StopSpinner(sp)

	if err != nil {
		var apiErr *api.APIError
		if isAPIError(err, &apiErr) {
			if Out.Mode == output.ModeJSON {
				return Out.LogJSON(map[string]any{
					"command": "yank",
					"status":  "error",
					"error":   apiErr.Code,
					"message": apiErr.Message,
				})
			}
			if apiErr.IsNotFound() {
				Out.LogError("%s@%s does not exist in the registry.", name, version)
			} else if apiErr.IsUnauthorized() {
				Out.LogError("Authentication failed. Run %s to re-authenticate.", output.Cyan("spm login"))
			} else {
				Out.LogError("Yank failed: %s", apiErr.Message)
			}
			return fmt.Errorf("yank failed: %s", apiErr.Message)
		}
		Out.LogError("Yank failed: %s", err)
		return fmt.Errorf("yank failed: %w", err)
	}

	if Out.Mode == output.ModeJSON {
		return Out.LogJSON(map[string]any{
			"command":   "yank",
			"status":    "success",
			"name":      resp.Name,
			"version":   resp.Version,
			"yanked":    resp.Yanked,
			"reason":    resp.Reason,
			"yanked_at": resp.YankedAt,
		})
	}

	Out.Log("%s Yanked %s@%s", output.Green(output.Icons["success"]),
		output.Cyan(name), version)
	if resp.Reason != "" {
		Out.Log("  Reason: %s", output.Dim(resp.Reason))
	}

	return nil
}
